Dockerfile

```diff
@@ -1,46 +1,36 @@
 FROM python:3.5.3-alpine
-MAINTAINER paul.traylor@linecorp.com
+LABEL maintainer=paul.traylor@linecorp.com
 
 ENV PROMETHEUS_VERSION 2.1.0
 ENV PROMETHEUS_DOWNLOAD_URL https://github.com/prometheus/prometheus/releases/download/v${PROMETHEUS_VERSION}/prometheus-${PROMETHEUS_VERSION}.linux-amd64.tar.gz
 
-RUN adduser -D -u 1000 promgen promgen
-RUN apk add --no-cache --update mariadb-dev build-base bash && \
-    rm -rf /var/cache/apk/*
-
-RUN set -ex; \
-        apk add --no-cache curl tar; \
-        curl -L -s $PROMETHEUS_DOWNLOAD_URL \
-		| tar -xz -C /usr/local/bin --strip-components=1 prometheus-${PROMETHEUS_VERSION}.linux-amd64/promtool; \
-        apk del curl tar; \
-	rm -rf /var/cache/apk;
-
 ENV PYTHONUNBUFFERED 1
 ENV PIP_NO_CACHE_DIR off
 
-RUN mkdir -p /etc/prometheus
-RUN mkdir -p /etc/promgen
-RUN mkdir -p /usr/src/app
-RUN chown promgen /etc/prometheus
-
 COPY docker/requirements.txt /tmp/requirements.txt
-RUN pip install -r /tmp/requirements.txt
-
 COPY setup.py /usr/src/app/setup.py
 COPY promgen /usr/src/app/promgen
 COPY promgen/tests/examples/promgen.yml /etc/promgen/promgen.yml
+COPY docker/docker-entrypoint.sh /
 
 WORKDIR /usr/src/app
-RUN pip install -e .
 
 ENV CONFIG_DIR=/etc/promgen
 
-USER promgen
```

Comment:
I have to admit I'm quite ignorant of docker best practices, but I thought it was generally discourated to run as root. I know Prometheus 2.0 was changed to not run as root ( https://github.com/prometheus/prometheus/issues/1637 ) though I think that also caused issues for some other things. Hmm.

Reply:
You are right, it is not as simple as I wrote in the description.

It is theoretically not possible for the docker root user to break out (and gain host root privileges). In production, containers are usually run as a non-root user to add an additional layer of security. And to prevent attackers from compromising the (ephemeral) container fs, of course.

Using `nobody` as the user to execute commands would be good for security and does not pose an issue for the Promgen Web App. The worker, however, needs write access to an external / shared volume.
Adding a new user to a docker image would require very specific configuration on the host system: Creating a new isolated user with the same uid and granting access only to the docker volume mount point.
By using root the "self-contained" aspect of the docker image is preserved. In more advanced or production deployments the user will (and should) most likely be changed to a custom user by the operator.

Reply:
I started reading up on the issue, as I'm not the most up-to-date either.

While most agree that it's best not to run as root, there is no consensus on what to do instead.
To me, it seems best to run under a random but static uid by default.

For Promgen the situation is slightly more complicated, as setting up write access to the external mount point is not optional but required for running the worker.

 * <https://github.com/prometheus/prometheus/issues/3441>
 * <https://github.com/minio/minio/pull/4877>
 * <https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user>